File: lib/src/models/model/gltf_model.dart
import 'package:playx_3d_scene/src/models/model/model.dart';

/// represents object of model that will be loaded from gltf file.
///
///glTF is a 3D file format maintained by the Khronos Group.
class GltfModel extends Model {
  /// Prefix path for gltf image assets to be added before image path.
  ///
  /// if the images path that in the gltf file different from the flutter asset path,
  /// consider adding prefix to the images path to be before the image.
  ///
  /// For example, if the image path in the gltf file is textures/texture.png
  /// and in assets the image path is assets/models/textures/texture.png
  /// you will need to add prefix to be 'assets/models/'.
  String prefix = "";

  /// postfix path for gltf image assets to be added after image path.
  ///
  /// if the images path that in the gltf file different from the flutter asset path,
  /// consider adding to the images path to be after the image.
  ///
  /// For example, if the image path in the gltf file is assets/textures/texture
  /// and in assets the image path is assets/textures/texture.png
  /// you will need to add prefix to be '.png'.
  String postfix = "";

  /// creates gltf model based on the  file asset path.
  GltfModel.asset(
    String path, {
    this.prefix = "",
    this.postfix = "",
    super.fallback,
    super.scale,
    super.centerPosition,
    super.animation,
  }) : super(assetPath: path);

  /// creates gltf model based on glb file url .
  /// currently supporting only .zip file format.
  GltfModel.url(
    String url, {
    this.prefix = "",
    this.postfix = "",
    super.fallback,
    super.scale,
    super.centerPosition,
    super.animation,
  }) : super(url: url);

  @override
  Map<String, dynamic> toJson() => {
        'assetPath': assetPath,
        'url': url,
        'pathPrefix': prefix,
        'pathPostfix': postfix,
        'fallback': fallback?.toJson(),
        'scale': scale,
        'centerPosition': centerPosition?.toJson(),
        'animation': animation?.toJson(),
        'isGlb': false,
      };

  @override
  String toString() {
    return 'GltfModel(assetPath: $assetPath, url: $url, prefix: $prefix, postfix: $postfix, fallback: $fallback, scale: $scale, centerPosition: $centerPosition, animation: $animation)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is GltfModel &&
        other.assetPath == assetPath &&
        other.url == url &&
        other.prefix == prefix &&
        other.postfix == postfix &&
        other.fallback == fallback &&
        other.scale == scale &&
        other.centerPosition == centerPosition &&
        other.animation == animation;
  }

  @override
  int get hashCode {
    return assetPath.hashCode ^
        url.hashCode ^
        prefix.hashCode ^
        postfix.hashCode ^
        fallback.hashCode ^
        scale.hashCode ^
        centerPosition.hashCode ^
        animation.hashCode;
  }
}

File: android/src/main/kotlin/io/sourcya/playx_3d_scene/core/model/gltf/model/GltfModel.kt
package io.sourcya.playx_3d_scene.core.model.gltf.model

import io.sourcya.playx_3d_scene.core.model.animation.model.Animation
import io.sourcya.playx_3d_scene.core.model.common.model.Model
import io.sourcya.playx_3d_scene.core.shape.common.model.Position
import io.sourcya.playx_3d_scene.core.shape.common.model.convertJsonToPosition
import io.sourcya.playx_3d_scene.utils.getMapValue

class GltfModel(
    assetPath: String? = null,
    url: String? = null,
    val pathPrefix: String = "",
    val pathPostfix: String = "",
    fallback: Model? = null,
    scale: Float? = null,
    centerPosition: Position?,
    animation: Animation? = null,
) : Model(assetPath,url, fallback, scale, centerPosition,animation){

    override fun toString(): String {
        return "GltfModel(pathPrefix='$pathPrefix', pathPostfix='$pathPostfix, animation=$animation, fallback=$fallback, position=$centerPosition, scale=$scale)')"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        if (!super.equals(other)) return false

        other as GltfModel

        if (pathPrefix != other.pathPrefix) return false
        if (pathPostfix != other.pathPostfix) return false

        return true
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + pathPrefix.hashCode()
        result = 31 * result + pathPostfix.hashCode()
        return result
    }


    companion object{


        fun fromMap(map: Map<String?, Any?>?): GltfModel? {
            if(map == null) return null
            val fallback = fromJson(getMapValue("fallback", map))
            val animation = Animation.fromJson(
                getMapValue("animation", map)
            )
            return GltfModel(
                assetPath = getMapValue("assetPath", map),
                url = getMapValue("url", map),
                pathPrefix = getMapValue("pathPrefix", map) ?: "",
                pathPostfix = getMapValue("pathPostfix", map) ?: "",
                fallback = fallback,
                scale = getMapValue("scale", map),
                centerPosition = convertJsonToPosition(getMapValue("centerPosition", map)),
                animation = animation,

            )

        }
    }

}